Index: include/llvm/ADT/Triple.h
===================================================================
--- include/llvm/ADT/Triple.h	(revision 171006)
+++ include/llvm/ADT/Triple.h	(working copy)
@@ -103,7 +103,9 @@
     NaCl,       // Native Client
     CNK,        // BG/P Compute-Node Kernel
     Bitrig,
-    AIX
+    AIX,
+      
+    TheOS
   };
   enum EnvironmentType {
     UnknownEnvironment,
Index: projects/sample/autoconf/config.sub
===================================================================
--- projects/sample/autoconf/config.sub	(revision 171006)
+++ projects/sample/autoconf/config.sub	(working copy)
@@ -1352,7 +1352,7 @@
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* | -theos*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
Index: lib/Support/Triple.cpp
===================================================================
--- lib/Support/Triple.cpp	(revision 171006)
+++ lib/Support/Triple.cpp	(working copy)
@@ -129,6 +129,7 @@
   case CNK: return "cnk";
   case Bitrig: return "bitrig";
   case AIX: return "aix";
+  case TheOS: return "theos";
   }
 
   llvm_unreachable("Invalid OSType");
@@ -276,6 +277,7 @@
     .StartsWith("cnk", Triple::CNK)
     .StartsWith("bitrig", Triple::Bitrig)
     .StartsWith("aix", Triple::AIX)
+    .StartsWith("theos", Triple::TheOS)
     .Default(Triple::UnknownOS);
 }
 
Index: autoconf/config.sub
===================================================================
--- autoconf/config.sub	(revision 171006)
+++ autoconf/config.sub	(working copy)
@@ -1352,7 +1352,7 @@
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* | -theos*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
